src/main/java/com/aws/greengrass/tes/HttpServerImpl.java

```diff
@@ -5,51 +5,203 @@
+        server.stop(TIME_TO_WAIT_BEFORE_SHUTDOWN_IN_SECONDS);
+    }
+
+    List<InetSocketAddress> getServerAddresses() {
+        return getServers().stream()
+                .map(HttpServer::getAddress)
+                .collect(Collectors.toList());
     }
 
     int getServerPort() {
-        return httpImpl.getAddress().getPort();
+        return getServers().stream()
+                .findFirst()
+                .map(HttpServer::getAddress)
+                .map(InetSocketAddress::getPort)
+                .orElse(PORT_UNKNOWN);
+    }
+
+    private List<HttpServer> getServers() {
+        return Stream.of(serverIPv4, serverIPv6)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    private void createServers(int port) throws IOException, InterruptedException {
+        if (!getServers().isEmpty()) {
+            logger.atInfo().kv("port", port).log("Skipping server creation, servers already running");
+            return;
+        }
+        // if port is system-picked, retry as we expect
+        // this to eventually succeed.
+        // otherwise, port is customer-configured,
+        // ensure we raise bind errors immediately.
+        if (port == 0) {
+            createServersWithRetry(port);
+        } else {
+            doCreateServers(port);
+        }
+    }
+
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.AvoidRethrowingException"})
+    private void createServersWithRetry(int port) throws IOException, InterruptedException {
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .maxAttempt(Integer.MAX_VALUE)
+                .initialRetryInterval(Duration.ofSeconds(1L))
+                .maxRetryInterval(Duration.ofMinutes(1L))
+                .retryableExceptions(Collections.singletonList(SocketException.class))
+                .build();
+
+        try {
+            RetryUtils.runWithRetry(retryConfig,
+                    () -> {
+                        doCreateServers(port);
+                        return null;
+                    },
+                    "create-http-servers",
+                    logger);
+        } catch (InterruptedException | IOException e) { // pass-through from RetryUtils
+            throw e;
+        } catch (Exception e) {
+            throw new RuntimeException("Unexpected exception during HTTP server creation", e);
+        }
+    }
+
+    /**
+     * Create ipv4 and ipv6 servers, both using the specified port.
+     *
+     * @param port chosen port for all created servers
+     * @throws IOException if an I/O error occurs during server creation
+     * @throws BindException if the chosen port is unavailable
+     */
+    private void doCreateServers(int port) throws IOException {
+        InetSocketAddress addrIPv4 = new InetSocketAddress(ADDR_IPV4, port);
+        try {
+            serverIPv4 = httpServerProvider.createHttpServer(addrIPv4, 0);
```

Comment:
Is it okay to block on this with infinite retries and never even attempt to create the ipv6 server or is the idea that this can only happen when the port is unavailable so the ipv6 server will also fail anyway? Would we instead benefit from configurable retries the default for which could be infinite retries?

Reply:
I think we should avoid having TES in a false half-state, where v4/v6 are both supported, but we only create one because the other fails due to port being unavailable.  So if v4 fails when we reasonably believe ipv4 is supported on the customer device, then it wouldn't make sense to try v6.  And inversely where v4 succeeds but v6 fails, we'd want to retry until we get both.

Configurable retries could be beneficial but I'm not sure if we'd be retrying much anyway.  I imagine the chance of a system picking a port that gets swiped before the v6 server is created is pretty low.  Maybe it would be better to fail more quickly to signal to the customer that there's some larger issue going on? (cc: @MikeDombo)

Reply:
Would we be able to do a lookup for `localhost` and get the address(es) for it? Instead of always trying to bind to ::1, we should only bind to ::1 and 127.0.01 if both are resolved for localhost address. This way, if a system only has ipv4 we should work fine, and the same if the system only has ipv6.